include/simics/dmllib.h

```diff
@@ -1785,6 +1788,124 @@ UNUSED static uint64 _identity_to_key(_identity_t id) {
         return (uint64)id.id << 32 | (uint64)id.encoded_index;
 }
 
+static attr_value_t
+_serialize_array_aux(const uint8 *data, size_t elem_size,
+                     const uint32 *dimsizes, uint32 dims,
+                     uint64 total_elem_count, _serializer_t serialize_elem) {
+    uint32 len = *dimsizes;
+
+    // Serialize the final dimension as bytes if serialize_elem is NULL
+    if (dims == 1 && !serialize_elem) {
+        return SIM_make_attr_data(len, data);
+    }
+
+    size_t children_elem_count = total_elem_count/len;
+
+    attr_value_t val = SIM_alloc_attr_list(len);
+    attr_value_t *items = SIM_attr_list(val);
+
+    for (uint32 i = 0; i < len; ++i) {
+        if (dims == 1) {
+            serialize_elem(&data[i*elem_size], &items[i]);
+        } else {
+            items[i] = _serialize_array_aux(
+                data + children_elem_count * elem_size * i, elem_size,
+                dimsizes + 1, dims - 1, children_elem_count, serialize_elem);
+        }
+    }
+    return val;
+}
+
+UNUSED static attr_value_t
+_serialize_array(const void *data, size_t elem_size,
+                 const uint32 *dimsizes, uint32 dims,
+                 _serializer_t serialize_elem) {
+    ASSERT(dims > 0);
+    size_t total_elem_count = 1;
+    for (uint32 i = 0; i < dims; ++i) {
+        total_elem_count *= dimsizes[i];
+    }
+    return _serialize_array_aux((const uint8 *)data, elem_size, dimsizes, dims,
+                                total_elem_count, serialize_elem);
+}
+
+
+static set_error_t
+_deserialize_array_aux(attr_value_t val, uint8 *data, size_t elem_size,
+                       const uint32 *dimsizes, uint32 dims,
+                       size_t total_elem_count,
+                       _deserializer_t deserialize_elem,
+                       bool elems_are_bytes) {
+    uint32 len = *dimsizes;
+
+    // Allow the final dimension to be represented as data if elems_are_bytes
+    bool data_allowed = elems_are_bytes && dims == 1;
+    if (data_allowed && SIM_attr_is_data(val)) {
+        if (unlikely(SIM_attr_data_size(val) != len)) {
+            SIM_c_attribute_error(
+                "Invalid serialized value of byte array: expected data of %u "
+                "bytes, got %u", len, SIM_attr_data_size(val));
+            return Sim_Set_Illegal_Value;
+        }
+        memcpy(data, SIM_attr_data(val), len);
+        return Sim_Set_Ok;
+    } else if (unlikely(!SIM_attr_is_list(val))) {
+        if (data_allowed) {
```

Comment:
I think this code is unreachable for attributes, since the declared attribute type captures the error earlier. But I suppose it's still useful for `after` event data. What will be the effect of SIM_attribute_error in this case?

Reply:
If `SIM_attribute_error` is called during deserialization of event data Simics will recognize it and checkpoint restoration will fail (and the error will be printed).